whalebird: 4.5.4 -> 4.6.0, add aarch64 support

uninsane opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

~~depends on #176181 . i'm not sure the right way to indicate that dependency, so i just included the commits from that PR in this branch.~~ (i've closed the referenced PR; this one will do both the version bump and new platform support)

uninsane committed:
whalebird: 4.5.4 -> 4.6.0; add aarch64 support

Whalebird distributes debs only for x86_64. but it distributes compiled .tar.bz2's
for x86_64 and aarch64. aarch64 support requires us to change the
build script to use these tarballs instead of the debs.

WolfangAukang commented:
I see you are already bumping whalebird with this PR. Why not close #176181?

WolfangAukang commented on pkgs/applications/misc/whalebird/default.nix:
```diff
@@ -22,7 +40,10 @@ stdenv.mkDerivation rec {
 
   unpackPhase = ''
-    dpkg-deb -x ${src} ./
+    mkdir -p ./opt
+    tar -xf ${src} -C ./opt
+    # remove the version/target suffix from the untar'd directory
+    mv ./opt/Whalebird-* ./opt/Whalebird
```
Is there an issue if you just use `opt` instead of `./opt`?

WolfangAukang reviewed (changes requested):
Thank you for your contribution. Here are some observations.

uninsane commented on pkgs/applications/misc/whalebird/default.nix:
```diff
@@ -1,17 +1,35 @@
+    desktopName = "Whalebird";
+    genericName = "An Electron based Mastodon client for Windows, Mac and Linux";
+    categories = [ "Network" ];
+    exec = "opt/Whalebird/whalebird";
```
:heavy_check_mark: 
`$out` isn't in scope here, so i applied the `@out@` placeholder + install-time `substitute --subst-var out` technique i see used elsewhere (e.g. in `pkgs/games/scummvm/games.nix`)

WolfangAukang replied:
I see. I did a wrong review here as you can point the package name and it should work. I will throw another review to fix the `makeDesktopItem` structure to the recommended style.

uninsane commented:
> Thank you for your contribution. Here are some observations.

thanks for the feedback (and for your patience)! i think i've addressed it all, and i'll make a note to fold related PRs together in the future.

WolfangAukang commented on pkgs/applications/misc/whalebird/default.nix:
```diff
@@ -1,28 +1,47 @@
 , nodePackages, alsa-lib, gtk3, libdbusmenu, libxshmfence, mesa, nss }:
 
+let
+  desktopItem = makeDesktopItem {
+    desktopName = "Whalebird";
+    genericName = "An Electron based Mastodon client for Windows, Mac and Linux";
+    categories = [ "Network" ];
+    exec = "@out@/bin/whalebird";
+    icon = "whalebird";
+    name = "whalebird";
+  };
```
```suggestion
  desktopItem = [
    (makeDesktopItem {
      desktopName = "Whalebird";
      comment = meta.description;
      categories = [ "Network" ];
      exec = "whalebird";
      icon = "whalebird";
      name = "whalebird";
    })
  ];
```

Generally, the `desktopItem` structure should be a list. Also, I've replaced the `genericName` part with `comment`, as the description kind of matches the previous content. You can still use `genericName`, but try to make into something shorter (like "Mastodon Client").

WolfangAukang replied:
FWIW, the `genericName`/`comment` part is just personal taste, but it would be helpful to avoid redundant stuff.

uninsane commented:
@WolfangAukang i applied your desktopItem suggestion. any more i can do to help the PR along?

WolfangAukang commented on pkgs/applications/misc/whalebird/default.nix:
```diff
@@ -1,28 +1,48 @@
-{ lib, stdenv, fetchurl, dpkg, autoPatchelfHook, makeWrapper, electron
+{ lib, stdenv, fetchurl, autoPatchelfHook, makeDesktopItem, makeWrapper, electron
```
```suggestion
{ lib, stdenv, fetchurl, autoPatchelfHook, makeDesktopItem, copyDesktopItems, makeWrapper, electron
```

This is to add the desktop items to the package.

WolfangAukang commented on pkgs/applications/misc/whalebird/default.nix:
```diff
@@ -1,28 +1,48 @@
   nativeBuildInputs = [
-    dpkg
     autoPatchelfHook
     makeWrapper
```
```suggestion
    makeWrapper
    copyDesktopItems
```

This is to add the desktop items to the package.

WolfangAukang reviewed (changes requested):
Only thing missing, as the desktop file is not being generated with the result. Other than that, the package works well and everything looks good.

Also, this is optional, but as you are adding a new platform (aarch64), I think it would be better to consider adding yourself to the maintainers in case there is an issue. If you want to do it, just throw a commit to add yourself as a maintainer by editing the maintainers file first, then add yourself below to the list of maintainers of the package.

SuperSandro2000 commented on pkgs/applications/misc/whalebird/default.nix:
```diff
@@ -1,28 +1,48 @@
+        url = downloads + "/Whalebird-${version}-linux-arm64.tar.bz2";
+        sha256 = "de0cdf7cbd6f0305100a2440e2559ddce0a5e4ad73a341874d6774e23dc76974";
+      };
+    }.${stdenv.system};
```
Missing or throw unsupported platform

uninsane committed:
maintainers: add colinsane

uninsane commented:
i've applied the feedback:
- added `copyDesktopItems`
- throws on unsupported systems (i think this should already be guarded by `platforms = [ "x86_64-linux" "aarch64-linux" ];` but i made it explicit)
- added myself as maintainer

SuperSandro2000 reviewed (commented):
Please squash the package related commits together.

uninsane commented:
> Please squash the package related commits together.

squashed into one commit for `maintainer-list.nix` and one commit for all the whalebird changes.
